### PR TITLE
[#70] Implement a helper Map implementation for ConfigSources to use

### PR DESCRIPTION
**Pull request by dmlloyd**

### CODE REVIEW

**Comment by dmlloyd** [comment]
Added tests.

File: implementation/pom.xml [hunk 1]
```diff
@@ -69,6 +69,50 @@
           <forkMode>once</forkMode>
         </configuration>
       </plugin>
+      <plugin>
```
Comment (kenfinnigan) [inline implementation/pom.xml]: Should something like this be moved into `smallrye-parent` so all projects could use it?

File: implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+ * This should not be used to implement {@link ConfigSource#getProperties()} on {@code ConfigSource}
+ * instances which do not override {@code getPropertyNames()}, as this will result in infinite recursion.
+ *
+ * @implNote The key set of the map is the result of calling {@link ConfigSource#getPropertyNames()}; the rest
+ * of the map operations are derived from this method and {@link ConfigSource#getValue(String)}.
+ * The values collection and entry set are instantiated lazily and cached.
+ * The implementation attempts to make no assumptions about the efficiency of the backing implementation and
+ * prefers the most direct access possible.
+ * <p>
+ * Some mutate operations are supported (specifically removals), where they can be performed on the backing key set, if
```
Comment (jmesnil) [inline implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java]: that's not specified in the spec (although it should be) but my interpretation of `getPropertyNames()` is that it should return an immutable set of names.
In general, `ConfigSource` should not be used to modify/update configuration. If some of these methods provide ambiguity about it, it should be properly specified in the spec or in the API.

I'll open issues in `microprofile-config` to clarify these use cases

File: implementation/pom.xml (hunk 1 above)
Comment (dmlloyd) [reply implementation/pom.xml]: I'd be okay with that I guess.

File: implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    public void clear() {
+        delegate.getPropertyNames().clear();
+    }
+
+    public String remove(final Object key) {
+        final String value = get(key);
+        //noinspection SuspiciousMethodCalls - it's OK in this case
+        return delegate.getPropertyNames().remove(key) ? value : null;
```
Comment (jmesnil) [inline implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java]: `remove` could throw a `UnsupportedOperationException` if the list is unmodifiable

File: implementation/pom.xml (hunk 1 above)
Comment (dmlloyd) [reply implementation/pom.xml]: But that means there'd have to be a release and so forth.

Comment (kenfinnigan) [reply implementation/pom.xml]: True, but pulling the trigger is pretty easy.

Other than needing a release of parent, any reason not to?

Comment (dmlloyd) [reply implementation/pom.xml]: No, sounds fine.  Is that something you can do?

Comment (kenfinnigan) [reply implementation/pom.xml]: Sure

File: implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java (hunk 2 above)
Comment (jmesnil) [reply implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java]: The API says:  https://github.com/eclipse/microprofile-config/blob/4611c9433675e34ee73bd1eba026e01631582772/api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java#L39

Comment (jmesnil) [reply implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java]: there is also https://github.com/eclipse/microprofile-config/issues/333 where it is discussed

**Comment by dmlloyd** [comment]
The idea was to "punt" on mutability.  If the underlying config source is mutable then the map is; if it isn't, then the map isn't.  This way it doesn't impose any policy one way or the other.

I can make it assume full immutability though if you want me to.

**Comment by jmesnil** [comment]
@dmlloyd I would prefer full immutability (the goal being then to propose this improvement to microprofile-config to update its API wrt `getProperties()/getPropertyNames()` ).

I don't like that the application could change the config sources.
First, the app does not "own" the config source, it may not be the only one that uses the config sources (eg config maps, DBs).
Then the config source itself may not be the same one that provides the properties depending on the stage (e.g. on dev, the prop would come from embedded property files, on prod, from a config map).

Having mutability raise expectation that the application knows the config source that always provides the property which is the opposite of the design behind the `Config` interface.

The whole microprofile-config API is really only about viewing the configuration. The update/modification of the configuraiton is out of scope of the specification.
However, this is not clearly enough stated in the spec and the use of mutable interface such as `Properties` leaves open the possibility to have mutable ConfigSources.

**Comment by dmlloyd** [comment]
OK I'll push an immutable version.

**Comment by dmlloyd** [comment]
OK pushed (also remembered to update the tests ;) )

**Comment by dmlloyd** [comment]
I think CI might have been confused by my rapid pushes.  Can we trigger a retest?

**Comment by dmlloyd** [comment]
I guess I'll just re-amend and push a new commit ID

**Comment by dmlloyd** [comment]
Hmm

**Comment by dmlloyd** [comment]
Ah it's the JavaDoc issue.  We need Ken's parent update.

**Comment by kenfinnigan** [comment]
v2 has been pushed to Sonatype, basically waiting for percolation to Central.

**Comment by kenfinnigan** [comment]
`2` is now in Central

**Comment by dmlloyd** [comment]
Thanks Ken, I'll update the PR.

**Comment by jmesnil** [comment]
@dmlloyd could you please rebase and I'll merge it?

**Pull request merged**